### ISSUE TITLE
apiserver request timeouts causing healthy node to be removed from LB's target pool

### ISSUE THREAD

**Issue by improbable-yifan** [kubernetes/kubernetes#48603]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

One of our etcd node had some networking issues that caused all writes to that node to time out for a minute or so.

```
Jul 05 02:27:36 **** etcd2[1767]: got unexpected response error (etcdserver: request timed out) [merged 52 repeated lines in 5.33s]
Jul 05 02:27:36 **** etcd2[1767]: etcdserver: request timed out, possibly due to connection lost [merged 2 repeated lines in 5.33s]
```

We see a bunch of extremely long node status update requests on the k8s API server that talks to the broken etcd node. 

```
I0705 02:27:36.414234       1 trace.go:61] Trace "Update /api/v1/nodes/****/status" (started 2017-07-05 02:27:06.412880726 +0000 UTC):
	[51.454µs] [51.454µs] About to convert to expected version
	[214.683µs] [147.831µs] Conversion done
	[227.295µs] [3.469µs] About to store object in database
	[30.000636968s] [30.000390408s] END
```

A lot of kubelet workers had issues updating their node status when they hit the bad API server due to request timeout (default 30s).

```
E0705 02:27:36.420298   21359 kubelet_node_status.go:302] Error updating node status, will retry: Timeout: request did not complete within allowed duration
```

Controller set all of those nodes’ status to be unhealthy as it has not observed a status update from the those nodes due to the request timeout (default 40s).

```
I0705 02:26:27.503832       1 controller_utils.go:285] Recording status change NodeNotReady event message for node ****
```

The controller then moved on to removing all those nodes from the target pool of our L3LBs.
Due the number of node affected it had to update around 9 target pools which took around 3 minutes from 02:27:00 - 02:30:00.

During this time all the node managed to successfully report their status and get recognised as ready again.

```
I0705 02:27:57.287101    4513 kubelet_node_status.go:358] Recording NodeReady event message for node ****
```

Those nodes are then added back into the L3LBs’ target pools by the controller but because there's only a single goroutine is setup (this is the default) to sync node state to LBs, it still hasn’t completed the jobs for removing those nodes from L3LBs targets. It took another 3 minutes to complete from 02:30:00 - 02:33:00, soon after it finished the previous removal requests.

**What you expected to happen**:

- request timeout to be way shorter than 30s for node status update as the default heartbeat interval is 10s.
- controller to not update the cloud providers target pools if the node has become healthy before it updates the L3LB’s target pools to remove the node.

**How to reproduce it (as minimally and precisely as possible)**:

- Make a request k8s API server to timeout requests and observe the cascading failure.

**Environment**:

Kubernetes version (use `kubectl version`):

```
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.7", GitCommit:"8eb75a5810cba92ccad845ca360cf924f2385881", GitTreeState:"clean", BuildDate:"2017-04-27T09:42:05Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```
Cloud provider or hardware configuration**: Google Cloud Platform
OS (e.g. from /etc/os-release): 

```
$ cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1325.1.0
VERSION_ID=1325.1.0
BUILD_ID=2017-02-23-0035
PRETTY_NAME="Container Linux by CoreOS 1325.1.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```

Kernel (e.g. `uname -a`): 

```
Linux <hostname> 4.9.9-coreos-r1 #1 SMP Thu Feb 23 00:27:53 UTC 2017 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux
```
Install tools:
Others:

**Comment by improbable-yifan** [comment]
@kubernetes/sig-api-machinery-*

**Comment by improbable-yifan** [comment]
@kubernetes/sig-api-machinery-

**Comment by lavalamp** [comment]
@dchen1107 @davidopp What sig owns the node controller?

**Comment by davidopp** [comment]
For the last half-year or so, no SIG has owned the NodeController. @gmarek has been the owner/maintainer out of the goodness of his heart, but this is not a tenable long-term situation. Finding a SIG to own it has been on my TODO list for many months but previous efforts were not successful. I will send email to some relevant SIG leads to see if we can find an owner.

**Comment by davidopp** [comment]
But I guess for this particular issue, we can hope @gmarek can take a quick look at it...

**Comment by gmarek** [comment]
I don't think it's strictly node controller issue. IIUC the problem is with Service (if that's what you mean by L3LB) and the fact that NC marks Pods running on Unreachable Nodes as 'Unknown'. I guess service controller removes such Pods from Endpoints and does all the magic around this (@improbable-yifan please correct me if that's not the case). I think that what @improbable-yifan suggested is actually correct course of action - we probably need to shorten our timeout for write requests (@kubernetes/sig-api-machinery-feature-requests) and fix whatever is reconfiguring LBs (@kubernetes/sig-network-feature-requests) so that all queued operation would be cancellable.

From the NC perspective we can think what we want to do with PodStatus for Pods having some tolerations running on notReady/unreachable Nodes once we go with taint-based-evictions to beta. @gyliu513 @k82cn

**Comment by gg7** [comment]
I also encountered the same issue today.

Environment:
* Kubernetes v1.6.6 (7 nodes, 3 of which masters)
* etcd: "gcr.io/google_containers/etcd:2.2.5", 3 servers (running on the master nodes)
* Ubuntu 16.04, 64 bit
* On-premise / bare-metal deployment, with custom deployment

At 13:46:43, master number 2 experienced a problem with its network driver:

    i40e 0000:02:00.0: TX driver issue detected, PF reset issued

This made it partitioned from the rest of the network (wouldn't even respond to ping), until it was power-cycled 35 minutes later.

At 13:46:44 a new etcd master was elected. At 13:46:52 our monitoring tools updated the k8s master A record (we use round-robin DNS). At 13:47:08 a new controller manager was elected. At 13:47:39-41 all master nodes were declared to be "NotReady", even though only master 2 was down.

What's interesting in my case is that the first "Error updating node status" message in the kubelet logs is from 14:02:15 (on both master 1 and 3), even though I use `--node-status-update-frequency=5s`. Any ideas why that's the case? Why did kubelet need 15 minutes to report its node status?

Anyway, Kubernetes is advertised as ["production-grade"](https://kubernetes.io/) and ["highly-available"](https://kubernetes.io/docs/admin/high-availability/). I have followed the documentation and have made everything redundant, yet disconnecting a single server broke my entire cluster for 15-20 minutes.

Is patching the source code the only workaround?
